### ISSUE TITLE
Support for manifest lists and digests

### ISSUE THREAD

**Issue by hassenius** [plexsystems/sinker#11]
This looks like a really interesting project. We're currently have some Jenkins jobs to sync specific images between registres using Skopeo, and this may be a nicer approach for us moving forward.
However, most of our images are multi-arch and inspired by Openshift 4 we reference everything by digest rather than tag.

Are multi-arch (manifest list) and digests something you currently support or plan to support?

**Comment by jpreese** [comment]
Thanks for trying the project, @hassenius! Digests are currently not supported, but something that can definitely be added.

**Comment by jpreese** [comment]
@hassenius for your use case with digests, what would be the expected tag when pushing the digest to the target repository?

**Comment by hassenius** [comment]
This is actually more of an interesting problem than one might think at first.

From the consumption side, we don't care what the tags are, however, if we overwrite existing tags the digest will be garbage collected and become invalid. 

So in our case we have different implementations, varying from using a set value (i.e. version number) with some conflict handling, to just using the digest value (i.e whatever the value after `sha256:`).

I _think_ as a general rule, those who are interested in digest, have less interest in tags, and can live with whatever happens.. ;)

**Comment by jpreese** [comment]
Put another way, what would be the expected behavior that you're looking for with digests? For example, it's possible to expose something along the lines of:

```yaml
target:
  host: myhost.com
sources:
- repository: myrepo
  digest: sha256:abc123
```

Where you have a digest that exists in `myrepo`, and you want to sync it to `myhost.com`. Is the expected behavior that you would push the resulting artifact that was pulled down with a tag of the digest? e.g. /myrepo:abc123

**Comment by hassenius** [comment]
Yes, that would be the perfect default behaviour, and fit our use case perfectly.

In addition it would be useful for those who might still want to use tags as additional metadata to allow deciding a tag, or default tag
So for example

```yaml
target:
  host: myhost.com
  default_tag: 1.0.0
sources:
- repository: myrepo
  digest: sha256:abc123
- repository: myrepo2
  digest: sha256:cde456
  tag: 1.0.0-abc
```

**Comment by jpreese** [comment]
Thanks for the clarification @hassenius, I think I understand what you're looking for and it all makes sense to do. I am a little hesitant about the `default_flag`. At the moment it seems like it could be confusing and I'd rather have users specify it, but that thought can change.

So for an initial implementation, given the following manifest:

```yaml
target:
  host: myhost.com
  default_tag: 1.0.0
sources:
- repository: myrepo
  digest: sha256:abc123
- repository: myrepo2
  digest: sha256:cde456
  tag: 1.0.0-abc
```

You would expect:
myrepo@sha256:abc123 to be pulled and then pushed to myhost.com/repo:abc123

myrepo2@sha256:cde456 to be pulled and then pushed to myhost.com/repo:cde456 AND be pushed to myhost.com/repo:1.0.0-abc

Correct?

**Comment by hassenius** [comment]
Thanks for engaging discussion @jpreese  In my initial thought I wasn't necessarily thinking to double push, but rather as a mechanism of collision handling, i.e. if two versions of the same image would prevent using the "top level" default tag.

So in other words, the order would be 
1. image tag
2. target.default_tag
3. sources[].digest value

Looking at your proposed manifest with a `.sources[].target` override field does look much cleaner and more intuitive though.

**Comment by jpreese** [comment]
No problem! In the interest of making progress on this, I'll add the digest feature first and we'll build off of that to address additional use cases.

I'll add a digest field that will pull the digest and then push it with a tag that matches the digest.

**Comment by jpreese** [comment]
@hassenius in your use case with digests, the expected behavior when using `sinker pull target` would still be to pull the specific digest, and not the tag correct? Does it matter? i.e.

syncing image@sha256:123 will

pull `image@sha256:123` from source and push it to `image:123`

`sinker pull target` could either:

pull the digest listed, or pull image:123, the _tag_ version of the digest.

My current thought is that when using digests, that tags are more or less ignored. The tooling will always pull the _digest_ (if not on the local machine) if the digest field is set, and always push the tag (no choice here) and _re-write a tag_ that already exists with that same SHA if one exists.

This gets into the conflict cases you were talking about I believe.

**Comment by jpreese** [comment]
@hassenius digests are now available in `v0.9.0`. Please give this a try and if there are additional cases that we need to consider for your use case, we'll address them in a new issue! Thanks for bringing up this issue.
